File: articles/superconductivity-superfluidity-and-macroscopic-quantum-order/r/ginzburg_landau_free_energy.R
# Ginzburg-Landau Free Energy
#
# Evaluates:
#
#   f(|Psi|) = alpha(T) |Psi|^2 + beta/2 |Psi|^4
#
# with alpha(T) = alpha_0 (T - Tc).

library(tidyverse)

article_dir <- "articles/superconductivity-superfluidity-and-macroscopic-quantum-order"

cases <- read_csv(file.path(article_dir, "data/gl_parameter_cases.csv"), show_col_types = FALSE)

amplitudes <- tibble(
  psi_amplitude = seq(0, 4, length.out = 300)
)

free_energy_table <- cases %>%
  rowwise() %>%
  mutate(temperature_k = list(seq(temperature_min_k, temperature_max_k, length.out = 80))) %>%
  unnest(temperature_k) %>%
  ungroup() %>%
  crossing(amplitudes) %>%
  mutate(
    alpha = alpha_0 * (temperature_k - critical_temperature_k),
    free_energy_density = alpha * psi_amplitude^2 + 0.5 * beta * psi_amplitude^4
  )

minimum_table <- free_energy_table %>%
  group_by(case_id, temperature_k) %>%
  slice_min(free_energy_density, n = 1, with_ties = FALSE) %>%
  ungroup() %>%
  mutate(
    analytic_equilibrium_amplitude =
      if_else(alpha < 0, sqrt(-alpha / beta), 0)
  )

write_csv(
  free_energy_table,
  file.path(article_dir, "data/computed_gl_free_energy_r.csv")
)

write_csv(
  minimum_table,
  file.path(article_dir, "data/computed_gl_minimum_table_r.csv")
)

print(minimum_table)

File: articles/superconductivity-superfluidity-and-macroscopic-quantum-order/python/ginzburg_landau_free_energy.py
"""
Ginzburg-Landau Free Energy

This workflow evaluates the uniform Ginzburg-Landau free-energy density:

    f = alpha(T) |Psi|^2 + beta/2 |Psi|^4

with:

    alpha(T) = alpha_0 (T - T_c)
"""

from pathlib import Path

import numpy as np
import pandas as pd


def run_case(case: pd.Series) -> pd.DataFrame:
    """
    Compute free-energy curves and equilibrium amplitude for one case.
    """
    temperatures = np.linspace(
        float(case["temperature_min_k"]),
        float(case["temperature_max_k"]),
        80,
    )
    amplitudes = np.linspace(0.0, 4.0, 300)

    rows = []

    for temperature in temperatures:
        alpha = float(case["alpha_0"]) * (temperature - float(case["critical_temperature_k"]))
        free_energy = alpha * amplitudes**2 + 0.5 * float(case["beta"]) * amplitudes**4

        min_index = int(np.argmin(free_energy))

        rows.append(
            {
                "case_id": case["case_id"],
                "temperature_k": temperature,
                "critical_temperature_k": float(case["critical_temperature_k"]),
                "alpha": alpha,
                "beta": float(case["beta"]),
                "equilibrium_amplitude_numeric": amplitudes[min_index],
                "minimum_free_energy_density": free_energy[min_index],
                "equilibrium_amplitude_analytic": np.sqrt(-alpha / float(case["beta"]))
                if alpha < 0
                else 0.0,
                "notes": case["notes"],
            }
        )

    return pd.DataFrame(rows)


def main() -> None:
    """
    Generate Ginzburg-Landau equilibrium amplitude tables.
    """
    article_dir = Path(__file__).resolve().parents[1]
    cases = pd.read_csv(article_dir / "data" / "gl_parameter_cases.csv")

    output = pd.concat([run_case(case) for _, case in cases.iterrows()], ignore_index=True)
    output.to_csv(article_dir / "data" / "computed_gl_equilibrium_amplitudes.csv", index=False)

    print("Ginzburg-Landau equilibrium amplitude summary:")
    print(output.groupby("case_id").head(6).round(8).to_string(index=False))


if __name__ == "__main__":
    main()
